int-set/src/bitset.rs

```diff
@@ -0,0 +1,402 @@
+                .insert_range(page_start, page_end);
+        }
+        self.mark_dirty();
+    }
+
+    /// Remove val from this set.
+    pub fn remove(&mut self, val: T) -> bool {
+        let val = val.into();
+        let page = self.page_for_mut(val);
+        let ret = page.remove(val);
+        self.mark_dirty();
+        ret
+    }
+
+    /// Returns true if val is a member of this set.
+    pub fn contains(&self, val: T) -> bool {
+        let val = val.into();
+        self.page_for(val)
+            .map(|page| page.contains(val))
+            .unwrap_or(false)
+    }
+}
+
+impl<T> BitSet<T> {
+    /// Remove all members from this set.
+    pub fn clear(&mut self) {
+        self.pages.clear();
+        self.page_map.clear();
+        self.mark_dirty();
+    }
+
+    /// Returns the number of members in this set.
+    pub fn len(&self) -> usize {
+        // TODO(garretrieger): keep track of len on the fly, rather than computing it. Leave a computation method
+        //                     for complex cases if needed.
+        if self.is_dirty() {
+            // this means we're stale and should recompute
+            let len = self.pages.iter().map(|val| val.len()).sum();
+            self.len.set(len);
+        }
+        self.len.get()
+    }
+
+    fn iter_pages(&self) -> impl Iterator<Item = (u32, &BitPage)> + '_ {
+        self.page_map
+            .iter()
+            .map(|info| (info.major_value, &self.pages[info.index as usize]))
+    }
+
+    fn iter_non_empty_pages(&self) -> impl Iterator<Item = (u32, &BitPage)> + '_ {
+        self.iter_pages().filter(|(_, page)| !page.is_empty())
+    }
+
+    fn mark_dirty(&mut self) {
+        self.len.set(usize::MAX);
+    }
+
+    fn is_dirty(&self) -> bool {
+        self.len.get() == usize::MAX
+    }
+
+    /// Return the major value (top 23 bits) of the page associated with value.
+    fn get_major_value(&self, value: u32) -> u32 {
+        return value >> PAGE_BITS_LOG_2;
+    }
+
+    fn major_start(&self, major: u32) -> u32 {
+        return major << PAGE_BITS_LOG_2;
+    }
+
+    /// Return a reference to the that 'value' resides in.
+    fn page_for(&self, value: u32) -> Option<&BitPage> {
+        let major_value = self.get_major_value(value);
+        self.page_map
+            .binary_search_by(|probe| probe.major_value.cmp(&major_value))
+            .ok()
+            .and_then(|info_idx| {
+                let real_idx = self.page_map[info_idx].index as usize;
+                self.pages.get(real_idx)
+            })
+    }
+
+    /// Return a mutable reference to the that 'value' resides in. Insert a new
+    /// page if it doesn't exist.
+    fn page_for_mut(&mut self, value: u32) -> &mut BitPage {
+        let major_value = self.get_major_value(value);
+        self.page_for_major_mut(major_value)
+    }
+
+    // Return a mutable reference to the page with major value equal to major_value.
+    // Inserts a new page if it doesn't exist.
```

Comment:
might as well make these comments into doc comments, even if the method is private; it means that at in our own use we will see these comments on hover in an IDE, for example


```suggestion
    /// Return a mutable reference to the page with the provided major value.
    ///
    /// Inserts a new page if it doesn't exist.
```